File: src/evals.rs
use std::fs;
use std::io;
use std::path::Path;

use serde::{Deserialize, Serialize};

use crate::clock::now_iso;
use crate::config::WorkContext;
use crate::git::{commit_diff, head_sha, worktree_diff, GitError};
use crate::lessons::append_candidate;
use crate::task::{read_artifact, read_intent, Task};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum EvalOutcome {
    Done,
    Blocked,
}

impl EvalOutcome {
    fn as_str(self) -> &'static str {
        match self {
            Self::Done => "done",
            Self::Blocked => "blocked",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EvalCase {
    pub id: String,
    pub ts: String,
    pub outcome: String,
    pub reason: Option<String>,
    pub worktree: String,
    pub base_commit: String,
    pub verify: Option<String>,
    pub spec: String,
    pub diff: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CorrectionEvalCase {
    pub id: String,
    pub ts: String,
    pub outcome: String,
    pub reason: String,
    pub note: Option<String>,
    pub worktree: String,
    pub base_commit: String,
    pub verify: Option<String>,
    pub spec: String,
    pub agent_attempt: String,
    pub human_fix: String,
}

pub fn capture_eval_case(
    ctx: &WorkContext,
    task: &Task,
    outcome: EvalOutcome,
    reason: Option<&str>,
) {
    if let Err(err) = capture_eval_case_at(ctx, task, outcome, reason, &now_iso(), None) {
        eprintln!("eval capture failed for {}: {err}", task.id);
    }
}

pub fn capture_correction(ctx: &WorkContext, task: &Task, note: &str) {
    let suffix = {
        let now = time::OffsetDateTime::now_utc();
        ((now.unix_timestamp() as i128 * 1000) + (now.nanosecond() as i128 / 1_000_000)).to_string()
    };
    let lesson = if note.trim().is_empty() {
        "see eval candidate"
    } else {
        note.trim()
    };
    if let Err(err) =
        capture_correction_eval_at(ctx, task, note, "manual", lesson, &now_iso(), &suffix)
    {
        eprintln!("correction capture failed for {}: {err}", task.id);
    }
}

pub fn capture_eval_case_at(
    ctx: &WorkContext,
    task: &Task,
    outcome: EvalOutcome,
    reason: Option<&str>,
    timestamp: &str,
    filename_suffix: Option<&str>,
) -> Result<Option<String>, CaptureEvalError> {
    if !ctx.config.capture_evals {
        return Ok(None);
    }
    let spec = read_intent(task)?;
    let committed = outcome == EvalOutcome::Done && task.meta.commit.is_some();
    let base_commit = if committed {
        format!("{}^", task.meta.commit.as_deref().unwrap_or_default())
    } else {
        head_sha(&ctx.root)?
    };
    let diff = if committed {
        commit_diff(&ctx.root, task.meta.commit.as_deref().unwrap_or_default())?
    } else {
        worktree_diff(&ctx.root)?
    };
    let record = EvalCase {
        id: task.id.clone(),
        ts: timestamp.to_string(),
        outcome: outcome.as_str().to_string(),
        reason: reason.map(ToOwned::to_owned),
        worktree: ctx.root.clone(),
        base_commit,
        verify: task.meta.verify.clone(),
        spec,
        diff,
    };
    let dir = format!("{}/eval-candidates", ctx.repo_state_dir);
    fs::create_dir_all(&dir)?;
    let suffix = filename_suffix.map(ToOwned::to_owned).unwrap_or_else(|| {
        let now = time::OffsetDateTime::now_utc();
        ((now.unix_timestamp() as i128 * 1000) + (now.nanosecond() as i128 / 1_000_000)).to_string()
    });
    let path = format!("{dir}/{}.{}.{}.json", task.id, outcome.as_str(), suffix);
    write_json(&path, &record)?;
    Ok(Some(path))
}

pub fn capture_correction_eval_at(
    ctx: &WorkContext,
    task: &Task,
    note: &str,
    category: &str,
    lesson: &str,
    timestamp: &str,
    filename_suffix: &str,
) -> Result<Option<String>, CaptureEvalError> {
    let intent = read_intent(task)?;
    let agent_attempt = read_artifact(task, "diff.patch")?.unwrap_or_default();
    let human_fix = worktree_diff(&ctx.root)?;
    let reason = task
        .meta
        .note
        .clone()
        .unwrap_or_else(|| "blocked".to_string());

    let path = if ctx.config.capture_evals {
        let record = CorrectionEvalCase {
            id: task.id.clone(),
            ts: timestamp.to_string(),
            outcome: "corrected".to_string(),
            reason,
            note: (!note.is_empty()).then(|| note.to_string()),
            worktree: ctx.root.clone(),
            base_commit: head_sha(&ctx.root)?,
            verify: task.meta.verify.clone(),
            spec: intent,
            agent_attempt,
            human_fix,
        };
        let dir = format!("{}/eval-candidates", ctx.repo_state_dir);
        fs::create_dir_all(&dir)?;
        let path = format!("{dir}/{}.corrected.{filename_suffix}.json", task.id);
        write_json(&path, &record)?;
        Some(path)
    } else {
        None
    };

    append_candidate(
        ctx,
        &format!(
            "correction - {} - [{}] {}",
            task.id,
            category,
            if lesson.is_empty() {
                "see eval candidate"
            } else {
                lesson
            }
        ),
    );
    Ok(path)
}

fn write_json(path: impl AsRef<Path>, value: &impl Serialize) -> io::Result<()> {
    fs::write(
        path,
        format!(
            "{}\n",
            serde_json::to_string_pretty(value).map_err(io::Error::other)?
        ),
    )
}

#[derive(Debug, thiserror::Error)]
pub enum CaptureEvalError {
    #[error(transparent)]
    Io(#[from] io::Error),
    #[error(transparent)]
    Git(#[from] GitError),
}

#[cfg(test)]
mod tests {
    use std::process::{Command, Stdio};

    use tempfile::TempDir;

    use crate::config::{Agent, AgentCli, AgentsConfig, AskConfig, Config, RoleAgents};
    use crate::task::{add_task, write_artifact, AddTaskOptions};

    use super::*;

    fn ctx(dir: &TempDir, capture_evals: bool) -> WorkContext {
        let root = dir.path().to_string_lossy().to_string();
        WorkContext {
            root: root.clone(),
            config: Config {
                capture_evals,
                agents: AgentsConfig::default(),
                ask: AskConfig::default(),
                ..Config::default()
            },
            state_dir: format!("{root}/state"),
            tasks_dir: format!("{root}/state/tasks"),
            plans_dir: None,
            agents: RoleAgents {
                planners: Vec::new(),
                implementer: Agent {
                    cli: AgentCli::Codex,
                    model: None,
                    provider: None,
                },
                reviewer: Agent {
                    cli: AgentCli::Claude,
                    model: None,
                    provider: None,
                },
                delivery: Agent {
                    cli: AgentCli::Claude,
                    model: None,
                    provider: None,
                },
            },
            ask_agent: Agent {
                cli: AgentCli::Claude,
                model: None,
                provider: None,
            },
            repo_state_dir: format!("{root}/repo-state"),
            metrics_path: format!("{root}/repo-state/metrics.db"),
        }
    }

    fn git(dir: &TempDir, args: &[&str]) {
        let status = Command::new("git")
            .arg("-C")
            .arg(dir.path())
            .args(args)
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status()
            .unwrap();
        assert!(status.success(), "git {args:?} failed");
    }

    fn init_repo(dir: &TempDir) {
        git(dir, &["init"]);
        git(dir, &["config", "user.email", "factory@example.com"]);
        git(dir, &["config", "user.name", "Factory"]);
        fs::write(dir.path().join("README.md"), "base\n").unwrap();
        git(dir, &["add", "README.md"]);
        git(dir, &["commit", "-m", "base"]);
    }

    #[test]
    fn capture_eval_writes_blocked_worktree_candidate() {
        let dir = TempDir::new().unwrap();
        init_repo(&dir);
        let ctx = ctx(&dir, true);
        let task = add_task(
            &ctx,
            "Fix the thing",
            Some("cargo test".to_string()),
            AddTaskOptions::default(),
        )
        .unwrap();
        fs::write(dir.path().join("README.md"), "changed\n").unwrap();

        let path = capture_eval_case_at(
            &ctx,
            &task,
            EvalOutcome::Blocked,
            Some("review failed"),
            "2026-06-22T12:00:00.000Z",
            Some("123"),
        )
        .unwrap()
        .unwrap();

        let text = fs::read_to_string(path).unwrap();
        assert!(text.contains(r#""outcome": "blocked""#));
        assert!(text.contains(r#""reason": "review failed""#));
        assert!(text.contains(r#""verify": "cargo test""#));
        assert!(text.contains("# git status"));
        assert!(text.contains("README.md"));
    }

    #[test]
    fn capture_eval_returns_none_when_disabled() {
        let dir = TempDir::new().unwrap();
        init_repo(&dir);
        let ctx = ctx(&dir, false);
        let task = add_task(&ctx, "No capture", None, AddTaskOptions::default()).unwrap();

        assert_eq!(
            capture_eval_case_at(
                &ctx,
                &task,
                EvalOutcome::Blocked,
                None,
                "2026-06-22T12:00:00.000Z",
                Some("123"),
            )
            .unwrap(),
            None
        );
    }

    #[test]
    fn capture_correction_writes_eval_and_lesson_candidate() {
        let dir = TempDir::new().unwrap();
        init_repo(&dir);
        let ctx = ctx(&dir, true);
        let task = add_task(&ctx, "Correct this", None, AddTaskOptions::default()).unwrap();
        write_artifact(&task, "diff.patch", "agent diff").unwrap();
        fs::write(dir.path().join("README.md"), "human fix\n").unwrap();

        let path = capture_correction_eval_at(
            &ctx,
            &task,
            "human note",
            "testing",
            "assert the right thing",
            "2026-06-22T12:00:00.000Z",
            "456",
        )
        .unwrap()
        .unwrap();

        let text = fs::read_to_string(path).unwrap();
        assert!(text.contains(r#""outcome": "corrected""#));
        assert!(text.contains(r#""agentAttempt": "agent diff""#));
        assert!(text.contains("human fix"));
        let candidates =
            fs::read_to_string(format!("{}/LESSONS.candidates.md", ctx.repo_state_dir)).unwrap();
        assert!(candidates.contains("[testing] assert the right thing"));
    }
}

File: src/evals.ts
import { mkdir } from 'node:fs/promises'
import { runAgent } from './agents.ts'
import type { WorkContext } from './config.ts'
import { commitDiff, headSha, worktreeDiff } from './git.ts'
import { createGuidanceFromDistillation } from './guidance.ts'
import { appendCandidate } from './lessons.ts'
import { log } from './log.ts'
import { correctionPrompt } from './prompts.ts'
import { readIntent, type Task } from './task.ts'

// Capture every terminal task as a reproducible eval candidate, so a regression
// set accrues from real use instead of being hand-authored. Each case is
// self-contained (the diff is stored inline, since worktrees are short-lived and
// their branches may be deleted): re-run by checking out `baseCommit`, running
// factory on `spec`, and scoring against `verify` + the reference `diff`.
// Best-effort — a capture failure is logged and never affects the task.

type EvalOutcome = 'done' | 'blocked'

export async function captureCorrectionGuidance(
  ctx: WorkContext,
  input: {
    taskId: string
    category: string
    lesson: string
    distillation: string
  }
): Promise<'created' | 'duplicate' | 'not-actionable' | 'invalid'> {
  return createGuidanceFromDistillation(ctx, {
    source: { kind: 'correction', taskId: input.taskId, detail: input.category },
    text: input.lesson,
    distillation: input.distillation,
  })
}

export async function captureEvalCase(
  ctx: WorkContext,
  task: Task,
  outcome: EvalOutcome,
  reason?: string
): Promise<void> {
  if (!ctx.config.captureEvals) {
    return
  }
  try {
    const spec = await readIntent(task)
    // done → the committed diff, based on its parent. blocked → the uncommitted
    // attempt, based on current HEAD (the work was never committed).
    const committed = outcome === 'done' && task.meta.commit
    const baseCommit = committed ? `${task.meta.commit}^` : await headSha(ctx.root)
    const diff = committed
      ? await commitDiff(ctx.root, task.meta.commit ?? '')
      : await worktreeDiff(ctx.root)

    const record = {
      id: task.id,
      ts: new Date().toISOString(),
      outcome,
      reason: reason ?? null,
      worktree: ctx.root,
      baseCommit,
      verify: task.meta.verify,
      spec,
      diff,
    }
    const dir = `${ctx.repoStateDir}/eval-candidates`
    await mkdir(dir, { recursive: true })
    await Bun.write(
      `${dir}/${task.id}.${outcome}.${Date.now()}.json`,
      JSON.stringify(record, null, 2)
    )
    log.info(`captured eval candidate: ${task.id} (${outcome})`)
  } catch (err) {
    log.warn(`eval capture failed for ${task.id}: ${err instanceof Error ? err.message : err}`)
  }
}

// Record a human takeover of a blocked task: pair the agent's failed attempt (the
// saved diff.patch) with the human's correction (the current worktree) and distill
// the highest-signal lesson there is — what the agent should have done, with the
// right answer in hand. Writes a paired eval candidate + a lesson candidate.
export async function captureCorrection(ctx: WorkContext, task: Task, note: string): Promise<void> {
  const intent = await readIntent(task)
  const attempt = Bun.file(`${task.dir}/diff.patch`)
  const agentAttempt = (await attempt.exists()) ? await attempt.text() : ''
  const humanFix = await worktreeDiff(ctx.root)
  const reason = task.meta.note ?? 'blocked'

  let category = 'other'
  let lesson = note
  let distillation: string | null = null
  try {
    const out = await runAgent(ctx.agents.reviewer, {
      root: ctx.root,
      prompt: correctionPrompt(intent, agentAttempt, humanFix, note, reason),
      access: 'read',
    })
    distillation = out.text
    category = /CATEGORY:\s*(\w+)/i.exec(out.text)?.[1]?.toLowerCase() ?? 'other'
    lesson = /LESSON:\s*(.+)/i.exec(out.text)?.[1]?.trim() ?? note
  } catch (err) {
    log.warn(
      `correction distill failed for ${task.id}: ${err instanceof Error ? err.message : err}`
    )
  }

  if (ctx.config.captureEvals) {
    try {
      const record = {
        id: task.id,
        ts: new Date().toISOString(),
        outcome: 'corrected',
        reason,
        note: note || null,
        worktree: ctx.root,
        baseCommit: await headSha(ctx.root),
        verify: task.meta.verify,
        spec: intent,
        agentAttempt,
        humanFix,
      }
      const dir = `${ctx.repoStateDir}/eval-candidates`
      await mkdir(dir, { recursive: true })
      await Bun.write(
        `${dir}/${task.id}.corrected.${Date.now()}.json`,
        JSON.stringify(record, null, 2)
      )
    } catch (err) {
      log.warn(
        `correction eval capture failed for ${task.id}: ${err instanceof Error ? err.message : err}`
      )
    }
  }

  await appendCandidate(
    ctx,
    `correction · ${task.id} · [${category}] ${lesson || 'see eval candidate'}`
  )
  if (distillation && lesson) {
    try {
      const captured = await captureCorrectionGuidance(ctx, {
        taskId: task.id,
        category,
        lesson,
        distillation,
      })
      if (captured === 'invalid') {
        log.warn(`correction guidance metadata invalid for ${task.id}; kept raw candidate only`)
      }
    } catch (err) {
      log.warn(
        `correction guidance capture failed for ${task.id}: ${
          err instanceof Error ? err.message : err
        }`
      )
    }
  }
  log.info(`recorded correction for ${task.id} (${category})`)
}
